src/protocols/login.js

```diff
@@ -0,0 +1,253 @@
+  const nonceVerified = nonceDB.searchAndDelete(jwsPayload.data.challenge);
+  if (nonceVerified === undefined) {
+    return false;
+  }
+
+  // check that jwsPayload.proofOfKSign.proofs.length <= 2
+  if (jwsPayload.proofOfKSign.proofs.length > 2) {
+    return false;
+  }
+
+  // check times iat <= current < exp
+  const date = new Date();
+  const current = Math.round((date).getTime() / 1000);
+  if (!((jwsHeader.iat <= current) && (current < jwsHeader.exp))) {
+    return false;
+  }
+
+  // check jwsPayload.ksign with jwsPayload.proofOfKSign.Leaf
+  // get ClaimAuthorizeKSign from jwsPayload.proofOfKSign.leaf
+  let entry = new Entry();
+  entry.fromHexadecimal(jwsPayload.proofOfKSign.leaf);
+  const claimAuthorizeKSign = claimUtils.newClaimFromEntry(entry);
+  const pubK = claimAuthorizeKSign.structure.pubKeyCompressed;
+  const addr = ethUtil.pubToAddress(pubK, true);
+  const addrHex = utils.bytesToHex(addr);
+  if (addrHex !== jwsPayload.ksign) {
+    return false;
+  }
+
+  // check that jwsPayload.form.proofOfEthName == jwsPayload.form.ethName == jwsHeader.iss
+  // get ClaimAssignName from jwsPayload.form.proofOfEthName.leaf
+  entry = new Entry();
+  entry.fromHexadecimal(jwsPayload.form.proofOfEthName.proofOfClaimAssignName.leaf);
+  const claimAssignName = claimUtils.newClaimFromEntry(entry);
+  const nameWithoutDomain = jwsPayload.form.proofOfEthName.name.split('@')[0];
+  // check jwsPayload.form.proofOfEthName.proofOfClaimAssignName.leaf {hashName} === hash(jwsPayload.form.proofOfEthName.name
+  if (utils.bytesToHex(claimAssignName.structure.hashName) !== utils.bytesToHex(utils.hashBytes(nameWithoutDomain).slice(1, 32))) {
+    return false;
+  }
+  // check claimAssignName.structure.id = jwsHeader.iss
+  if (utils.bytesToHex(claimAssignName.structure.id) !== jwsHeader.iss) {
+    return false;
+  }
+
+  // check verify signature with jwsPayload.ksign
+  // as verifying a signature is cheaper than verifying a merkle tree proof, first we verify signature with ksign
+  const header64 = Buffer.from(JSON.stringify(jwsHeader)).toString('base64');
+  const payload64 = Buffer.from(JSON.stringify(jwsPayload)).toString('base64');
+  const dataSigned = header64 + '.' + payload64;
+  const message = ethUtil.toBuffer(dataSigned);
+  const msgHash = ethUtil.hashPersonalMessage(message);
+  const sigHex = utils.bytesToHex(signatureBuffer);
+  if (!utils.verifySignature(utils.bytesToHex(msgHash), sigHex, jwsPayload.ksign)) { // mHex, sigHex, addressHex
+    return false;
+  }
+
+  // TODO AFTER MILESTONE verify identity address from counterfactual
+  // TODO AFTER MILESTONE check counterfactual address from jwsPayload.identity, address == jwsHeader.iss
+  // TODO AFTER MILESTONE check jwsPayload.identity.relay == hardcoded relay address
+
+  const relayAddr = '0xe0fbce58cfaa72812103f003adce3f284fe5fc7c';
+
+  // verify proofOfEthName
+  if (!proofs.verifyProofClaimFull(jwsPayload.form.proofOfEthName.proofOfClaimAssignName, relayAddr)) {
```

Comment:
TODO: proofOfEthName should be the proofOfClaim:
from: {
  ethName: str
  proofOfEthName: proofOfClaim
}